### PR TITLE
Enhancement for ENV specs

### PR DESCRIPTION
**Pull request by BurdetteLamar**

### CODE REVIEW

File: core/env/keys_spec.rb [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 describe "ENV.keys" do
 
-  it "returns all the keys" do
-    ENV.keys.sort.should == ENV.to_hash.keys.sort
+  it "returns an array the keys" do
```
Comment (eregon) [inline core/env/keys_spec.rb]: `of` the keys

File: core/env/each_value_spec.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@
       ENV.clear
       ENV["1"] = "3"
       ENV["2"] = "4"
-      ENV.each_value { |v| e << v }
+        ENV.each_value { |v| e << v }.should equal(ENV)
```
Comment (eregon) [inline core/env/each_value_spec.rb]: Unneeded change of indentation

File: core/env/shared/to_hash.rb [hunk 3]
```diff
@@ -17,6 +24,9 @@
 
   it "duplicates the ENV when converting to a Hash" do
     h = ENV.send(@method)
-    h.should_not equal ENV
```
Comment (eregon) [inline core/env/shared/to_hash.rb]: This is a useful check, could you keep it?

**Comment by BurdetteLamar** [comment]
Thanks, eregon, for a thorough review!

File: core/env/each_value_spec.rb (hunk 2 above)
Comment (BurdetteLamar) [inline core/env/each_value_spec.rb]: I don't understand this.  There's no extra space in line 13 over at https://github.com/BurdetteLamar/spec/blob/master/core/env/each_value_spec.rb.

Comment (eregon) [reply core/env/each_value_spec.rb]: It seems fine now, maybe it was a tab or a bug of GitHub.

File: core/env/index_spec.rb [hunk 4]
```diff
@@ -3,4 +3,10 @@
 
 describe "ENV.index" do
   it_behaves_like :env_key, :index
+
+  it "warns aboout deprecation" do
```
Comment (eregon) [inline core/env/index_spec.rb]: about*

File: core/env/replace_spec.rb [hunk 5]
```diff
@@ -1,15 +1,52 @@
...
-    ENV["baz"].should == "bam"
-    ENV.delete "baz"
+    ENV.replace("foo" => "0", "bar" => "1").should equal(ENV)
+    ENV.size.should == 2
+    ENV["foo"].should == "0"
+    ENV["bar"].should == "1"
+  end
+
+  it "raises TypeError if the argument is not a Hash" do
+    -> { ENV.replace("foo" => "0", "bar" => "1").should raise_error(TypeError, "no implicit conversion of Object into Hash") }
```
Comment (eregon) [inline core/env/replace_spec.rb]: The replace value is a Hash here.
And `should raise_error` should be after the end of the lambda (`}`)

File: core/env/replace_spec.rb [hunk 6]
```diff
@@ -1,15 +1,52 @@
...
+    ENV.to_hash.should == @orig
+  end
+
+  it "raises TypeError if a key is not a String" do
+    -> { ENV.replace(Object.new => "0").should raise_error(TypeError, "no implicit conversion of Object into String") }
```
Comment (eregon) [inline core/env/replace_spec.rb]: Same here and below, the `.should raise_error` should be after the `}`.
I wonder why the CI doesn't fail though.

File: core/env/shared/update.rb [hunk 7]
```diff
@@ -19,19 +19,42 @@
...
+  # end
+
+  # BUG: This fails because the block's return value becomes the new name.
+  # it "does not use the block's return value as the value for a new name" do
+  #   ENV.delete("bar")
+  #   ENV.send @method, {"foo" => "0"}
+  #   ENV.send(@method, "bar" => "1") { |key, old, new| "Should not use this value" }
+  #   ENV["foo"].should == "0"
+  #   ENV["bar"].should == "1"
+  # end
```
Comment (eregon) [inline core/env/shared/update.rb]: Could you report those to https://bugs.ruby-lang.org/ ?

File: core/env/replace_spec.rb (hunk 6 above)
Comment (eregon) [reply core/env/replace_spec.rb]: Ah, it's because this just create a big lambda literal, but actually never calls it.
And the spec still "pass" because of the expectation below which passes.

File: core/env/index_spec.rb (hunk 4 above)
Comment (BurdetteLamar) [reply core/env/index_spec.rb]: Will fix.

File: core/env/replace_spec.rb (hunk 5 above)
Comment (BurdetteLamar) [reply core/env/replace_spec.rb]: K

File: core/env/replace_spec.rb (hunk 6 above)
Comment (BurdetteLamar) [reply core/env/replace_spec.rb]: K

File: core/env/shared/update.rb (hunk 7 above)
Comment (BurdetteLamar) [reply core/env/shared/update.rb]: Done (a while back): https://bugs.ruby-lang.org/issues/16192.

File: core/env/methods_spec.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+ruby_version_is "2.5" do
+  describe "ENV.class" do
+    ENV.should be_an_instance_of(Object)
+  end
+
+  describe "ENV.methods" do
+    it "returns an array of methods" do
+      methods = ENV.methods false
+      expected_methods = [:[], :[]=, :assoc, :clear, :delete, :delete_if, :each, :each_key, :each_pair, :each_value, :empty?, :fetch, :has_key?, :has_value?, :include?, :index, :inspect, :invert, :keep_if, :key, :key?, :keys, :length, :member?, :rassoc, :rehash, :reject, :reject!, :replace, :select, :select!, :shift, :size, :store, :to_a, :to_h, :to_hash, :to_s, :update, :value?, :values, :values_at]
```
Comment (eregon) [inline core/env/methods_spec.rb]: Listing all methods like this is fairly unusual for specs, I am not sure it's a good idea here.
Could you extract that commit in a separate PR?

In general, each method should have specs and passing those already guarantees all methods are implemented.
Having extra methods should probably not be forbidden.

File: core/env/shared/update.rb (hunk 7 above)
Comment (eregon) [inline core/env/shared/update.rb]: If we know this will be the behavior after these bugs are fixed, then it would be good to use `ruby_version_is "2.7" do`.

Comment (eregon) [reply core/env/shared/update.rb]: In any case we should link to the relevant MRI bug for each of these specs.

File: core/env/methods_spec.rb (hunk 8 above)
Comment (BurdetteLamar) [reply core/env/methods_spec.rb]: Sure, will separate.  I was thinking that this would prompt us to spec a new method.

File: core/env/replace_spec.rb [hunk 9]
```diff
@@ -1,15 +1,52 @@
...
+    ENV.to_hash.should == @orig
+  end
+
+  it "raises TypeError if a key is not a String" do
+    -> { ENV.replace(Object.new => "0") }.should raise_error(TypeError, "no implicit conversion of Object into String")
+    ENV.to_hash.should == @orig
+  end
+
+  it "raises TypeError if a value is not a String" do
+    -> { ENV.replace("foo" => Object.new).should raise_error(TypeError, "no implicit conversion of Object into String") }
```
Comment (eregon) [inline core/env/replace_spec.rb]: There were still a few issues here, I fixed them in 13a40d29a3a66f1b0dd8a8b6167d2b2d49f630a5.
